### DISCUSSION TITLE
Recommendations for handling a user object or other object that is needed throughout the app

### DISCUSSION THREAD

**Discussion by grangej** [pointfreeco/swift-composable-architecture#578] [Q&A]
I am looking for some best practices on how to handle user object that is mutable / updatable (such as profile changes, etc).

I have an app that has multiple tabs and other nested structure and I am curious how to handle pushing update of a user object throughout the app? 

This is what I have so far
- AppState holds on to the user object 
- I have UserStore that is responsible for updating caching the user object. 
- I have views "pull" from the UserStore onAppear 

What is a good way to have AppState push the user object to all nested views? In vanilla SwiftUI this would be a perfect case for an environment object. How can I replicate global state like environment object in TCA ?

**Comment by mluisbrown** [comment]
> How can I replicate global state like environment object in TCA ?

I asked a similar question a little while back and the [answer](https://github.com/pointfreeco/swift-composable-architecture/discussions/505#discussioncomment-647373) is that you can use the TCA `Environment` for this.

**Reply by grangej** [reply]
This is kinda like what I have where I have a UserStore that I pass in the environment, but the issue is the environment does not have the ability to push out changes and update the state of the view (unless I'm missing something?) 

So for example:

- 3 tabs 
- one tab has the ability to add points to the user object (available points) 
- another tab reads these points to determine UI 
- and clicking on a button in another tab might push a new view that also reads from the user object. 

Now imagine that I have the 3rd view pushed and then I switch to the other tab and manipulate the state of the user object, how does the 2 other views get updated? I understand in this simplistic example that I could just reach into the sub view states and "update the user" in the state objects but now imagine that there are further nested views and or more tabs. This becomes increasingly cumbersome to maintain and a huge amount of boilerplate code. How would you recommend telling all the views that care about the user object to update when the user object is mutated? 

----

Right now what I have again is a UserStore in the environment and whenever a view comes on screen I "ask" for the user object and update my local state. This seems odd and smells funny, but perhaps it's just a new way of doing things?

Thank you in advance!

**Reply by maximkrouk** [reply]
You may simply have a UserInfoClient in Environment even without making it composable and retrieve/update user data in lower-level modules as you need it, I think it's a bit more convenient than sharing such a small shared piece of data in an AppState, maybe it's a little less architectural, but it work for me. You may update your local state every time before switching tabs (if that logic is composable in your app).

Also, probably you may try to investigate how pointfree does it in the isowords.

**Comment by mbrandonw** [comment]
Transferring this to a discussion since it’s not an issue with the library.
